core/src/main/scala/cats/Defer.scala

```diff
@@ -72,6 +72,29 @@ trait Defer[F[_]] extends Serializable {
     lazy val res: F[A] = fn(defer(res))
     res
   }
+
+  /**
+   * Useful when you want a recursive function that returns F where
+   * F[_]: Defer. Examples include IO, Eval, or transformers such
+   * as EitherT or OptionT.
+   * 
+   * example:
+   *
+   * val sumTo: Int => Eval[Int] =
+   *   Defer[Eval].recursiveFn[Int, Int] { recur =>
+   *     
+   *     { i =>
+   *       if (i > 0) recur(i - 1).map(_ + i)
+   *       else Eval.now(0)
+   *     }
+   *   }
+   */
+  def recursiveFn[A, B](fn: (A => F[B]) => (A => F[B])): A => F[B] =
+    new Function1[A, F[B]] { self =>
+      lazy val loopFn: A => F[B] = fn(self)
```

Comment:
we make this a lazy val so it is only evaluated once *but* only inside the `defer` below so we want until the first time it is actually need.

Reply:
but now that I write that down, I can't see a reason why you would need this. You should only apply `self` inside the recursion, but that means only when `loopFn` is called, which we don't do until we return the value, so it should be safe.